src/ui/ui_settings/public/ui_settings_api.js

```diff
@@ -0,0 +1,48 @@
+import { sendRequest } from './send_request';
+
+export function createUiSettingsApi() {
+  let unsavedChanges = {};
+  let pendingCallbacks = [];
+  let timeout = null;
+
+  async function send() {
+    const changes = unsavedChanges;
+    const callbacks = pendingCallbacks;
+
+    pendingCallbacks = [];
+    unsavedChanges = {};
+
+    try {
+      const resp = await sendRequest({
+        method: 'POST',
+        path: '/api/kibana/settings',
+        body: { changes },
+      });
+
+      callbacks.forEach(cb => cb(null, resp));
```

Comment:
I think this part is slightly wrong.

a) we never remove any callbacks from `pendingCallbacks` ever again, which means they will stay registered forever, which doesn't actually cause a huge harm, since it's anyway just promise resolving methods, which doesn't matter if we call them again - except some small memory leak.
b) This might erroneously call a handler in the following scenario:
* Call `batchSet` (multiple) times.
* Wait 200ms.
* The request will be started.
* Call now `batchSet` while the request is still running.
* This last listener will now be resolved as soon as the first request returned, that actually didn't contain the changes for that listener.

Since these timing issues can usually cause very weird and hard to find issues, I wonder if we should have that "implicit" `batchSet` method that works with timeouts at all. What would happen if the user happens to press refresh (e.g. because someone implements it in the settings page, as requested by you in #4451 :D) some of the changes you thought that have happened are actually lost.

Maybe we should rather expose that as an actual API that either takes an array of changes, or a chaining API (`uiService.batch().set('foo', 'bar').set('foo2', 42).end()`) instead. That way we wouldn't have this timing issues within this API anymore, but would leave it to the consumer, to take care about canceling or sending batches, before page reloads, etc. happens.

Reply:
> a) we never remove any callbacks from pendingCallbacks ever again

we capture the current value of `pendingCallbacks` as the local `callbacks` binding and reassign the `pendingCallbacks` binding to an empty array each time the `send()` function is called. This should mean that only the `send()` closure has access to the callbacks registered before the call to send() and once it's complete the memory will be garbage collected.

> b) ... This last listener will now be resolved as soon as the first request returned, that actually didn't contain the changes for that listener.

Once `send()` has been called the `pendingCallbacks` array is replaced with an empty array, so the last call to `batchSet()` will push the callback into its own (empty) array and will only be called back once send runs again (not for the first response).

> Maybe we should rather expose that as an actual API that either takes an array of changes, or a chaining API (uiService.batch().set('foo', 'bar').set('foo2', 42).end()) instead. 

I'd be fine with another API, the only really important thing is that calls to `config.set()` synchronously effect the return value of `config.get()` and synchronously emit change events. Best I can tell we don't actually call `config.set()` multiple time anywhere, so I'll just get rid of the delay.

Reply:
:woman_facepalming: how the heck did I oversee the local variable. Sorry for the noise. Regarding the timing issue I'll let you the final decision. Maybe that also doesn't need to be in this PR and could go into a separate PR (if we need the API at all).